### PR TITLE
Feature/common sgd optimizers

### PR DESCRIPTION
**Pull request by VIVelev**
Optimizers implemented:
 - SGD
 - Momentum
 -  RMSprop
 - Adam

See [here](http://ruder.io/optimizing-gradient-descent/index.html) for more information about the optimizers.

--

Unit Tests were written for minimizing:
 - a simple quadratic function (basic) 
 - a matrix computed MSE Loss (matrix)

See `tests/test_optim/test_optimizers.py`.

### CODE REVIEW

File: nujo/optim/optimizers.py [hunk 1]
```diff
@@ -1,100 +1,138 @@
...
+
+    An iterative method for optimizing an objective function.
+
+    Parameters:
+    -----------
+    params : list of ndarray(s), the parameters which to update
+    lr : float, the learning rate
+
+    '''
     def __init__(self, params, lr=0.001):
```
Comment (ManiacMaxo) [inline nujo/optim/optimizers.py:32]: Type hinting

Comment (ManiacMaxo) [reply nujo/optim/optimizers.py:32]: on the whole file

File: tests/test_autodiff/test_functions.py [hunk 2]
```diff
@@ -163,4 +163,4 @@ def get_tensors():
     return A, B
 
 
-# ===================================================================================================
+# ====================================================================================================
```
Comment (ManiacMaxo) [inline tests/test_autodiff/test_functions.py:166]: Why did you extend the line breaks everywhere with one character??

Comment (VIVelev) [reply tests/test_autodiff/test_functions.py:166]: For consistency Maniac

**Pull request merged**